crates/sui-core/src/authority.rs

```diff
@@ -1559,7 +1550,7 @@ impl AuthorityState {
         });
 
         prometheus_registry
-            .register(Box::new(ModuleCacheGauge::new(&state.module_cache)))
+            .register(Box::new(ModuleCacheGauge::new(epoch_store.module_cache())))
```

Comment:
This won't work, unfortunately -- after the first epoch, the gauge will start reporting zero on the metric, because the cache that it was initialised with went away (it only holds a weak reference to it).

The two options are:

- Move the `ModuleCacheGauge` into the per-epoch store, and take care to [`unregister`](https://docs.rs/prometheus/latest/prometheus/fn.unregister.html) the old gauge and register the new one on each epoch boundary.
- Keep track of the size by instrumenting the `get_module` function on `ResolverWrapper`, which obviates the need for the special gauge -- this would involve adding a metric to `AuthorityMetrics`, registering it there, and then handing an accessor to that into `ResolverWrapper`. The new metric would probably need to be a `Counter` rather than a `Gauge` (i.e. you change from measuring the absolute value, to measuring the change in value).

The second option is better (I think it's the one we talked about before), but it might be a little more work.

Reply:
you are right we talked about that already. 
I'll hook up into the resolver and we'll go from there

Reply:
fixed, take a look please